readme.md
[![Run PHPUnit and PHPCS](https://github.com/trepmal/mini-loops/actions/workflows/integrate.yml/badge.svg)](https://github.com/trepmal/mini-loops/actions/workflows/integrate.yml)
[![Deploy to WordPress.org](https://github.com/trepmal/mini-loops/actions/workflows/deploy.yml/badge.svg)](https://github.com/trepmal/mini-loops/actions/workflows/deploy.yml)

# Mini Loops

Tested up to: 5.7  
License: GPLv2 or later

Widget for mini post loops.

Show most recent posts, posts from categories, and more. Can be displayed via widgets, shortcodes, and template tags.

This is a new release, please report bugs to trepmal (at) gmail (dot) com before leaving a poor review. Thanks

## Installation

1. Download the zip file and extract the contents.
2. Upload the 'mini-loops' folder to your plugin directory (default: wp-content/plugins/).
3. Activate the plugin through the 'plugins' page in WP.
4. See 'Appearance'->'Widgets' to place it on your sidebar. Set the settings.

## Frequently Asked Questions

### How can I exclude categories?

List its ID as a negative number.

### My site broke after customizing the excerpt. What happened?

Did you set the strip_tags attribute to false? (`[excerpt strip_tags=0]`). If tags haven't been stripped, it may have cut off the text before an HTML tag was properly closed, thus breaking the page. Setting strip_tags to false should only be used if you're carefully managing the excerpt's output.

### Why can't I add an ID to the element in 'Before Items'?

There are limitations on the tag or tag/attribute combiniations allowed by WordPress. However, these can be overcome with a few lines of code.

To allow an additional tag, use this:

```
add_filter( "admin_init", "allowed_tags" );
function allowed_tags() {
	global $allowedposttags;
	$allowedposttags["video"] = array();
}
```

To allow additional attributes for any tag, use the following code:

```
add_filter( "admin_init", "allowed_tags" );
function allowed_tags() {
	global $allowedposttags;
	$allowedposttags["video"]["src"] = array();
	$allowedposttags["video"]["type"] = array();
	$allowedposttags["video"]["poster"] = array();
}
```

If you're using a custom theme, you can add this code the its `functions.php` file. Otherwise, you may want to create your own plugin to hold these modifications.

See [this post](http://justintadlock.com/archives/2011/02/02/creating-a-custom-functions-plugin-for-end-users) for instructions on creating your own plugin.

## Screenshots

Widget options:

![Widget Options](.wordpress-org/screenshot-1.png)

More under [Sample Formats](https://github.com/trepmal/mini-loops#sample-item-formats)

## Other Notes

Explanation of options:

### Widget Display

**Title:** Your recent posts widget's title on your sidebar. Default: Recent Posts
`[miniloop title="Recent Posts"] get_miniloops( array('title' => 'Recent Posts' ) );`

**Title URL:** The page the title should link to. Default: none
`[miniloop title_url="/blog/"] get_miniloops( array('title_url' => '/blog/'' ) );`

### Building the Query

**Number of Posts:** Number of posts to be displayed. Default: 3
`[miniloop number_posts=3] get_miniloops( array('number_posts' => 3 ) );`

**Post Offset:** Number of posts to skip before displaying the list. Default: 0
`[miniloop post_offset=0] get_miniloops( array('post_offset' => 0 ) );`

**Maximum Age:** Don't show posts more than X days old Default: 0
`[miniloop maximum_age=0] get_miniloops( array('maximum_age' => 0 ) );`

**Post Author:** Get author's post. By ID. Default: none
`[miniloop post_author=0] get_miniloops( array('post_author' => 0 ) );`

**Post Type:** Post type to display. Default: post
`[miniloop post_type=post] get_miniloops( array('post_type' => 'post' ) );`

**Post Status:** Post status to display. Primarily useful to show upcoming (future) posts. But be creative! Default: publish
`[miniloop post_status=publish] get_miniloops( array('post_status' => 'publish' ) );`

**Order By:** What order the posts should be displayed in. Default: date
`[miniloop order_by=date] get_miniloops( array('order_by' => 'date' ) );`

**Order:** Ascending (good for order by title) or Descending (good for order by date) Default: DESC
`[miniloop order=DESC] get_miniloops( array('order' => 'DESC' ) );`

**Ordering by the values of `some_key`** (works well if the values of `some_key` are strings/words):
`[miniloop order_by=meta_value order_meta_key=some_key] get_miniloops( array('order_by' => 'meta_value', 'order_meta_key' => 'some_key' ) );`

**Ordering by the numeric values of `some_key`** (works well if the values of `some_key` are numbers/integers):
`[miniloop order_by=meta_value_num order_meta_key=some_key] get_miniloops( array('order_by' => 'meta_value_num', 'order_meta_key' => 'some_key' ) );`

**Show posts in reverse order?** Perhaps you want the 3 most recent posts, but you want the oldest of those to be displayed first. If so, check this. Default: 0
`[miniloop reverse_order=0] get_miniloops( array('reverse_order' => 0 ) );`

**Shuffle post order?** Shuffle the order of the posts matching your query. Perhaps showing the 5 most recent posts in randomized order. Default: 0
`[miniloop shuffle_order=0] get_miniloops( array('shuffle_order' => 0 ) );`

**Ignore sticky posts?** Treat sticky posts as normal posts. I recommend ignoring, or the number of posts displayed may be inconsistent. Default: 1
`[miniloop ignore_sticky=1] get_miniloops( array('ignore_sticky' => 1 ) );`

**Exclude sticky posts?** Don't show sticky posts at all. Default: 0
`[miniloop exclude_sticky=0] get_miniloops( array('exclude_sticky' => 0 ) );`

**Only sticky posts?** Show only sticky posts. Default: 0
`[miniloop only_sticky=0] get_miniloops( array('only_sticky' => 0 ) );`

**If viewing a single post, exclude it?** If viewing a single post, remove it from the widget display. Default: 1
`[miniloop exclude_current=1] get_miniloops( array('exclude_current' => 1 ) );`

**Get posts from current category (if archive)?** If viewing an archive, only show posts from the same category. Default: 0
`[miniloop current_category=1] get_miniloops( array('current_category' => 1 ) );`

**Get posts from first category (if single)?** If viewing a single post, only show posts from the first category. Default: none
`[miniloop current_single_category=1] get_miniloops( array('current_single_category' => 1 ) );`

**Get posts from current author (if single or archive)?** Show more posts from the current author. Default: none
`[miniloop current_author=1] get_miniloops( array('current_author' => 1 ) );`

**Categories:** Comma separated list of category IDs to pull from. Use negative ID numbers to exclude a category. Default: none
`[miniloop categories="1,8,13"] get_miniloops( array('categories' => '1,8,13' ) );`

**Tags:** Comma separated list of tag IDs to pull from. Use negative ID numbers to exclude a tag. Default: none
`[miniloop tags="15,40,88"] get_miniloops( array('tags' => '15,40,88' ) );`

**Custom Taxonomies:** A clunky way to support custom taxonomies. Default: none. To include terms 5, 6, 9 from taxonomy "Genre" do this:
`[miniloop tax="genre=5,6,9"] get_miniloops( array('tax' => 'genre=5,6,9' ) );`

**Custom Fields:** For listing posts that have certain meta data. Default: none. To list posts that have a custom field 'favorite_color' with a value of 'blue' do this:
`[miniloop custom_fields="favorite_color=blue"] get_miniloops( array('custom_fields' => 'favorite_color=blue' ) );`

**Exclude Posts:** A comma separated list of post IDs to exclude. Default: none
`[miniloop exclude="15,200,1032"] get_miniloops( array('exclude' => '15,200,1032' ) );`

### Display

**Before Items:** Text/HTML to insert before the post list. Default: `<ul>`
`[miniloop before_items="<div>"] get_miniloops( array('before_items' => '<div>' ) );`

**After Items:** Text/HTML to insert after the post list. Default: `</ul>`
`[miniloop after_items="</div>"] get_miniloops( array('after_items' => '</div>' ) );`

**Item Format:**
HTML and shortcodes to format each item

### Shortcodes

- `[ml_title]`
- `[ml_url]`
- `[ml_excerpt]` Attributes: length (100), wlength (0), after ('...'), space_between (0), after_link (1), custom (0), strip_tags (1), strip_shortcodes (1)
  - length = excerpt length in characters (0 for none, -1 for full length)
  - wlength = excerpt length in words
  - after = what to show after the excerpt
  - space_between = force space between excerpt and 'after'
  - after_link = make the 'after' link to the post
  - custom = 1 to default/customized excerpts, 0 to trim by lentgh
  - strip_tags = 1 to strip HTML tags, 0 to keep. **CAREFUL:** it is not generally recommended to keep the tags. Character excerpts may break tags, and thus break an entire page's layout.
  - strip_shortcodes = 1 to strip shortcodes, 0 to keep
  - up_to_more = 1 to get everything up to the `<!--more-->` tag (the 'more' text), if it exists, otherwise use char/word limit excerpt. 0 use char/word limited excerpt
  - after_with_more = (with up_to_more) 1 to use the 'after' text with the 'more' text, 0 to add nothing after the 'more' text
- `[ml_content]`
- `[ml_comment_count]`
- `[ml_author]`
- `[ml_author_link]`
- `[ml_author_avatar]` Attributes: size (92), default (''), alt (false)
  - size = avatar size, in pixels
  - default = default if no gravatar. See Settings > Discussion
  - alt = alt text
- `[ml_field]` Attributes: name, single (1), separator (', '), reverse (0)
  - name = custom field name
  - single = 1 get single value, 0 get all values matching name
  - separator = string to separate each value
  - reverse = 0 default order, 1 reverse display order
- `[ml_taxonomy]` Attributes: taxonomy, separator (', '), link (0), justone (0), reverse (0)
  - taxonomy = taxonomy slug
  - separator = string to separate each term
  - link = 1 to link categories to their archive page, 0 for no links
  - justone = 1 to show just first category, 0 to show all
  - reverse = 0 default order, 1 reverse display order
- `[ml_tax]` Alias to `[ml_taxonomy]`
- `[ml_category]` Shortcut for `[tax taxonomy=category]`
- `[ml_tag]` Shortcut for `[tax taxonomy=post_tag]`
- `[ml_date]` Attributes: format ('F j, Y')
  - format = PHP-style date format
- `[ml_class]` Attributes: class
  - class = classes to display in addition to the traditional post classes
- `[ml_image]` Attributes: from, cfname, class, width (50), height (50), crop, fallback
  - from (options: thumb, attached, customfield, first)
    - _from_ 'thumb' post thumbnail/featured image `[ml_image from=thumb]`
    - _from_ 'attached' first attached image `[ml_image from=attached]`
    - _from_ 'customfield' get from custom field `[ml_image from=customfield]`
    - _from_ 'first' first image in post `[ml_image from=first]`
  - since 1.1.3, you can pass a comma-separated list to use as fallbacks. `[ml_image from="thumb,first"]`
  - cfname = custom field to use if from=customfield `[ml_image from=customfield cfname=thumbnail]`
  - class = class for image
  - width = width of image
  - height = height of image
  - crop = 1 to crop, 0 to scale (not implemented yet)
  - fallback = URL of image to use if 'from' doesn't return anything
  - cache = set to 'clear' to generate new thumbnails. It is not recommended that you leave this option on. `[ml_image from=thumb cache=clear]`

Inside of Item Format, shortcodes can be used without the `ml_` prefix.

### Sample Item Formats

Format 1:

![format 1](.wordpress-org/screenshot-2.png)

(before: `<ul>` after: `</ul>`)

```
<li class="[class]"><p><a href="[url]">[image from=customfield cfname=image width=50 height=50 class=alignright fallback='http://dummyimage.com/50'][title]</a><br> [excerpt wlength=30 space_between=1 after="..." after_link=1]<br> By [author] on [date format="n/j/y"]</p><br></li>
```

Format 2:

![format 2](.wordpress-org/screenshot-3.png)

(before: `<ul>` after: `</ul>`)

```
<li class="[class]"><p>[date format="F j, Y"]<br /><a href="[url]"> [image from=customfield cfname=image width=380 height=200 class=aligncenter fallback='http://placekitten.com/380/200']</a> [excerpt length=90 space_between=1 after="..." after_link=1]</p></li>
```

Format 3:

![format 3](.wordpress-org/screenshot-4.png)

(before: -- after: --)

```

CHANGELOG.md
## Changelog

### Version 1.4.2

- Fix: misc php errors

### Version 1.4.1

- Fix: Corrects sloppy shortcode fixes.
- Fix: Mini-Mini-Loops widget default settings.

### Version 1.4

- Long-overdue cleanup.

### Version 1.3.1

- Fix: Widget title links

### Version 1.3

- Fix: Widget warnings since 4.3
- New: 'crop' parameter for the image tag
- New: Allow any shortcodes in item format
- New: Automatically clear thumbnail cache on image change. props @om4james
- Fix: Use widget_title filter

### Version 1.2

- Fix: undefined index notice if zero posts match query
- Fix: markup errors in widget. Corrects save issue regarding order
- Fix: reset postdata instead of query
- New: BETA - use `[ba_archive before='' after='']` shortcode to insert an author/taxonomy link. For use with before|after_items fields.
- Fix: removed deprecated function for thumbnail creation
- New: fallback 'from' options for [image]. [image from="thumb,first"] - If no post thumbnail (featured image), get first image from post
- New: [author_avatar] with optional parameters [author_avatar size=92 default='' alt=0]
- New: Multisite - Show posts from sister-sites (on same network). REQUIRES ADD ON: https://gist.github.com/trepmal/5073067
- New: Ajax-paging - View prev/next set of posts in widget. REQUIRES ADD ON: https://gist.github.com/trepmal/5073756

### Version 1.1.2

- Fix: imbalanced tags if zero posts match query

### Version 1.1

- New: Changed Before/After Items inputs to textareas for easier modifying if there is a lot of markup.
- New: Filters for the Before/After Items content. `miniloops_{before|after}_items_format` $query arguments passed to it. See source for more details.
- New: "Mini Mini Loops" widget. Simplified 'Recent Posts' widget for typical usage - only 3 options.
- New: Filter for altering query. `miniloops_query` See source for more details.
- New: Post author field.
- New: Get posts from current author (if viewing single post or author archive).
- New: 'length' 'before' 'after' parameters for [title] and [url] shortcode. Length is number of characters, processed prior to 'before' and 'after'.
- New: Maximum age field. Only show posts from last X number of days. Thanks bluey80.
- New: Meta value (alpha and numerical) ordering.
- New: [post_type] [post_type_archive_link]. Great for Before/After Item formats.
- New: 'Any' option for post type.
- New: Alt text support for thumbnails.
- Fix: Missed marking some strings for translation.
- Fix: Markup mixup for some selected options.
- General code clean up and improvements, such as full path used in include().

### Version 1.0.1

- Fix: Multiple tag bug. Only first was being recognized, now correctly accepts all. Thanks Ozias.

### Version 1.0

- New: Exclude sticky posts option.
- New: Get posts from first current category (if single).

### Version 0.9

- Fix: Prevents error from being displayed if image can't be resized.
- New: Improved support for multisite use.
- New: Improved [miniloop] shortcode. Editor tries to hard to "fix" the user-provided item format. Now you can save the format in a custom field. See Other Notes.

### Version 0.8 (2011.10.31)

- Added French Translation files. (Thanks [@maoneetag](http://twitter.com/maoneetag))

### Version 0.8

- New: [ml_field] shortcode, [ml_category] shortcode, [ml_tag] shortcode, [ml_taxonomy] shortcode.
- Fix: strip slashes from widget title.

### Version 0.7

- New: more excerpt options (use automated/custom excerpts rather than trim by length, option to bypass tag/shortcode stripping). Please report issues.
- Fix: stipping slashes for before/after item during output.

### Version 0.6

- New (sorta): shortcode option `[miniloop]` (see Other Notes for usage). Why "sorta"? shortcode has existed the whole time, I only just now added some docs.
- New: get posts from current category (if archive).
- New: custom taxonomies, if only 1 ID given, and it's negative, treat it like "NOT IN".
- Fix: added missing echo in instructions.
- General code optimization.

### Version 0.5

- New: get only sticky posts.
- New: shuffle order.
- New: [ml_author_link] shortcode.
- New: [ml_comment_count] shortcode.
- New: thumbnail cropping for local images.
- New: Ready for localization.

### Version 0.4

- New image option: get first attached image (great for galleries!).
- Bug fix: broken image if no fallback is set.

### Version 0.3

- Bug fix: post status works now.

### Version 0.2

- Improvements: hide title, link title, exclude current single post options.

### Version 0.1

- Initial release version.


form.php
<?php if ( ! defined( 'ABSPATH' ) ) die( '-1' ); ?>
		<p style="width:40%;float:left;">
			<label for="<?php echo esc_attr( $this->get_field_id( 'title' ) ); ?>"><?php esc_html_e( 'Title:', 'mini-loops' );?>
				<input class="widefat" id="<?php echo esc_attr( $this->get_field_id( 'title' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('title' ) ); ?>" type="text" value="<?php echo esc_attr( $title ); ?>" />
			</label>
		</p>
		<p style="width:40%;margin-left:2%;float:left;">
			<label for="<?php echo esc_attr( $this->get_field_id( 'title_url' )  ); ?>"><?php esc_html_e( 'Title URL:', 'mini-loops' );?>
				<input class="widefat" id="<?php echo esc_attr( $this->get_field_id( 'title_url' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('title_url' ) ); ?>" type="text" value="<?php echo esc_attr( $title_url ); ?>" />
			</label>
		</p>
		<p style="width:15%;margin-left:2%;float:left;padding-top:20px;height:20px;">
			<input type="checkbox" class="checkbox" id="<?php echo esc_attr( $this->get_field_id('hide_title' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('hide_title' ) ); ?>"<?php checked( $hide_title ); ?> />
			<label for="<?php echo esc_attr( $this->get_field_id('hide_title' ) ); ?>"><?php esc_html_e('Hide Title?', 'mini-loops' );?></label>
		</p>

		<hr style="clear:both;"/>

	<div style="width:48%;float:left;">

		<h3><?php esc_html_e( 'Query', 'mini-loops' ); ?></h3>
		<p style="width:48%;float:left;">
			<label for="<?php echo esc_attr( $this->get_field_id('number_posts' ) ); ?>"><?php esc_html_e('Number of Posts:', 'mini-loops' );?>
				<input class="widefat" id="<?php echo esc_attr( $this->get_field_id('number_posts' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('number_posts' ) ); ?>" type="number" value="<?php echo esc_attr( $number_posts ); ?>" />
			</label>
		</p>
		<p style="width:48%;float:right;">
			<label for="<?php echo esc_attr( $this->get_field_id('post_offset' ) ); ?>"><?php esc_html_e('Posts Offset:', 'mini-loops' );?>
				<input class="widefat" id="<?php echo esc_attr( $this->get_field_id('post_offset' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('post_offset' ) ); ?>" type="number" value="<?php echo esc_attr( $post_offset ); ?>" />
			</label>
		</p>
		<p style="width:48%;float:left;">
			<label for="<?php echo esc_attr( $this->get_field_id('maximum_age' ) ); ?>"><?php esc_html_e('Maximum Age:', 'mini-loops' );?>
				<input class="widefat" id="<?php echo esc_attr( $this->get_field_id('maximum_age' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('maximum_age' ) ); ?>" type="number" value="<?php echo esc_attr( $maximum_age ); ?>" />
				<small>(<?php esc_html_e('Only posts less than X days old.', 'mini-loops');?>) </small>
			</label>
		</p>
		<p style="width:48%;float:right;">
			<label for="<?php echo esc_attr( $this->get_field_id('post_author' ) ); ?>"><?php esc_html_e('Author:', 'mini-loops');?>
				<input class="widefat" id="<?php echo esc_attr( $this->get_field_id('post_author' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('post_author' ) ); ?>" type="text" value="<?php echo esc_attr( $post_author ); ?>" /><br />
				<small>(<?php esc_html_e('IDs, comma-separated.', 'mini-loops');?>) </small>
			</label>
		</p>
		<p>
		<p style="width:48%;float:left;clear:left;">
			<label for="<?php echo esc_attr( $this->get_field_id('post_type' ) ); ?>"><?php esc_html_e('Post Type:', 'mini-loops' );?>
				<select class="widefat" id="<?php echo esc_attr( $this->get_field_id('post_type' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('post_type' ) ); ?>">
				<?php
					echo "<option value='any'" . selected( 'any', $post_type, false ) . ">Any</option>";
					$pts = get_post_types( array( 'public' => true ), 'objects' );
					foreach ( $pts as $slug => $obj ) {
						printf(
							'<option value="%s" %s>%s</option>',
							esc_attr( $slug ),
							selected( $slug, $post_type, false ),
							esc_html( $obj->labels->name )
						);
					}
				?>
				</select>
			</label>
		</p>
		<p style="width:48%;float:right;">
			<label for="<?php echo esc_attr( $this->get_field_id( 'post_status' ) ); ?>"><?php esc_html_e('Post Status:', 'mini-loops' );?>
				<select class="widefat" id="<?php echo esc_attr( $this->get_field_id('post_status' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('post_status' ) ); ?>">
				<?php
					echo "<option value='any'" . selected( 'any', $post_status, false ) . ">Any</option>";
					if ( ! function_exists( 'get_available_post_statuses' ) ) {
						require_once ABSPATH . '/wp-admin/includes/post.php';
					}
					$pss = get_available_post_statuses();
					foreach ( $pss as $k => $v ) {
						printf(
							'<option value="%s" %s>%s</option>',
							esc_attr( $v ),
							selected( $v, $post_status, false ),
							esc_html( $v )
						);
					}
				?>
				</select>
			</label>
		</p>
		<p style="width:48%;float:left;">
			<label for="<?php echo esc_attr( $this->get_field_id('order_by' ) ); ?>"><?php esc_html_e('Order by:', 'mini-loops' );?>
				<select class="widefat" id="<?php echo esc_attr( $this->get_field_id('order_by' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('order_by' ) ); ?>">
				<?php
					$obs = array( 'id' => 'ID', 'author' => __('Author', 'mini-loops'), 'title' => __('Title', 'mini-loops'), 'date' => __('Date', 'mini-loops'), 'modified' => __('Last-modified Date', 'mini-loops'), 'parent', __('Parent ID', 'mini-loops'), 'rand' => __('Random', 'mini-loops'), 'comment_count' => __('Comment Count', 'mini-loops'), 'menu_order' => __('Menu Order', 'mini-loops'), 'meta_value' => __('Meta Value*', 'mini-loops'), 'meta_value_num' => __('Meta Value Numerical*', 'mini-loops') );
					foreach ( $obs as $k => $v ) {
						printf(
							'<option value="%s" %s>%s</option>',
							esc_attr( $k ),
							selected( $k, $order_by, false ),
							esc_html( $k )
						);
					}
				?>
				</select>
			</label>
		</p>
		<p style="width:48%;float:right;">
			<label for="<?php echo esc_attr( $this->get_field_id('order' ) ); ?>"><?php esc_html_e('Order:', 'mini-loops' );?>
				<select class="widefat" id="<?php echo esc_attr( $this->get_field_id('order' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('order' ) ); ?>">
				<?php
					$obs = array( 'ASC', 'DESC' );
					foreach ( $obs as $v ) {
						printf(
							'<option value="%s" %s>%s</option>',
							esc_attr( $v ),
							selected( $v, $order, false ),
							esc_html( $v )
						);
					}
				?>
				</select>
				<small>(<?php esc_html_e('ABC vs ZYX', 'mini-loops');?>)</small>
			</label>
		</p>
		<p style="clear:both;">
			<label for="<?php echo esc_attr( $this->get_field_id('order_meta_key' ) ); ?>"><?php esc_html_e('*Meta key required for meta value ordering', 'mini-loops');?></label>
			<input type="text" class="widefat" id="<?php echo esc_attr( $this->get_field_id('order_meta_key' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('order_meta_key' ) ); ?>" value="<?php echo esc_attr( $order_meta_key ); ?>" />
		</p>
		<p style="clear:both;">
			<input type="checkbox" class="checkbox" id="<?php echo esc_attr( $this->get_field_id('reverse_order' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('reverse_order' ) ); ?>"<?php checked( $reverse_order ); ?> />
			<label for="<?php echo esc_attr( $this->get_field_id('reverse_order' ) ); ?>"><?php esc_html_e('Show posts in reverse order?', 'mini-loops');?></label>
			<small>(<?php esc_html_e('ABC vs CBA', 'mini-loops');?>)</small>
		</p>
		<p>
			<input type="checkbox" class="checkbox" id="<?php echo esc_attr( $this->get_field_id('shuffle_order' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('shuffle_order' ) ); ?>"<?php checked( $shuffle_order ); ?> />
			<label for="<?php echo esc_attr( $this->get_field_id('shuffle_order' ) ); ?>"><?php esc_html_e('Shuffle post order?', 'mini-loops');?></label>
			<small>(<?php esc_html_e('ABC vs BCA', 'mini-loops');?>)</small>
		</p>
		<p style="width:48%;float:left;">
			<input type="checkbox" class="checkbox" id="<?php echo esc_attr( $this->get_field_id('ignore_sticky' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('ignore_sticky' ) ); ?>"<?php checked( $ignore_sticky ); ?> />
			<label for="<?php echo esc_attr( $this->get_field_id('ignore_sticky' ) ); ?>"><?php esc_html_e('Unstick sticky posts?', 'mini-loops');?></label>
		</p>
		<p style="width:48%;float:right;">
			<input type="checkbox" class="checkbox" id="<?php echo esc_attr( $this->get_field_id('only_sticky' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('only_sticky' ) ); ?>"<?php checked( $only_sticky ); ?> />
			<label for="<?php echo esc_attr( $this->get_field_id('only_sticky' ) ); ?>"><?php esc_html_e('Only sticky posts?', 'mini-loops');?></label>
		</p>
		<p style="width:48%;float:left;">
			<input type="checkbox" class="checkbox" id="<?php echo esc_attr( $this->get_field_id('exclude_sticky' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('exclude_sticky' ) ); ?>"<?php checked( $exclude_sticky ); ?> />
			<label for="<?php echo esc_attr( $this->get_field_id('exclude_sticky' ) ); ?>"><?php esc_html_e('Exclude sticky posts?', 'mini-loops');?></label>
		</p>
		<p style="clear:both;"><small><?php esc_html_e( '"Only" will take precedence if both sticky options are checked.', 'mini-loops' ); ?></small></p>
		<p>
			<input type="checkbox" class="checkbox" id="<?php echo esc_attr( $this->get_field_id('exclude_current' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('exclude_current' ) ); ?>"<?php checked( $exclude_current ); ?> />
			<label for="<?php echo esc_attr( $this->get_field_id('exclude_current' ) ); ?>"><?php esc_html_e('If viewing a single post, exclude it?', 'mini-loops');?></label>
		</p>
		<p style="clear:both;"><small><?php esc_html_e( 'This option is ignored if "Only sticky" is checked.', 'mini-loops' ); ?></small></p>
		<p>
			<input type="checkbox" class="checkbox" id="<?php echo esc_attr( $this->get_field_id('current_category' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('current_category' ) ); ?>"<?php checked( $current_category ); ?> />
			<label for="<?php echo esc_attr( $this->get_field_id('current_category' ) ); ?>"><?php esc_html_e('Get posts from current category (if archive)?', 'mini-loops');?></label>
		</p>
		<p>
			<input type="checkbox" class="checkbox" id="<?php echo esc_attr( $this->get_field_id('current_single_category' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('current_single_category' ) ); ?>"<?php checked( $current_single_category ); ?> />
			<label for="<?php echo esc_attr( $this->get_field_id('current_single_category' ) ); ?>"><?php esc_html_e('Get posts from first category (if single)?', 'mini-loops');?></label>
		</p>
		<p>
			<input type="checkbox" class="checkbox" id="<?php echo esc_attr( $this->get_field_id('current_author' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('current_author' ) ); ?>"<?php checked( $current_author ); ?> />
			<label for="<?php echo esc_attr( $this->get_field_id('current_author' ) ); ?>"><?php esc_html_e('Get posts from current author (if single or archive)?', 'mini-loops');?></label>
		</p>

	</div>

	<div style="width:48%;float:right;">

		<p style="width:48%;float:left;">
			<label for="<?php echo esc_attr( $this->get_field_id( 'categories' ) ); ?>"><?php esc_html_e('Categories:', 'mini-loops');?>
				<input class="widefat" id="<?php echo esc_attr( $this->get_field_id('categories' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('categories' ) ); ?>" type="text" value="<?php echo esc_attr( $categories ); ?>" /><br />
				<small>(<?php esc_html_e('IDs, comma-separated.', 'mini-loops');?>) </small>
			</label>
		</p>
		<p style="width:48%;float:right;">
			<label for="<?php echo esc_attr( $this->get_field_id( 'tags' ) ); ?>"><?php esc_html_e('Tags:', 'mini-loops');?>
				<input class="widefat" id="<?php echo esc_attr( $this->get_field_id('tags' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('tags' ) ); ?>" type="text" value="<?php echo esc_attr( $tags ); ?>" /><br />
				<small>(<?php esc_html_e('IDs, comma-separated.', 'mini-loops');?>) </small>
			</label>
		</p>
			<label for="<?php echo esc_attr( $this->get_field_id( 'tax' ) ); ?>"><?php esc_html_e('Custom Taxonomies:', 'mini-loops');?>
				<input class="widefat" id="<?php echo esc_attr( $this->get_field_id('tax' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('tax' ) ); ?>" type="text" value="<?php echo esc_attr( $tax ); ?>" /><br />
				<small>(<?php esc_html_e('Ex: category=1,2,4&amp;post_tag=6,12', 'mini-loops');?><br />
						<?php
						$taxonomy_names = get_taxonomies( array( 'public' => true ) );
						printf(
							esc_html('Available: %s', 'mini-loops'),
							implode( ', ', array_map( 'esc_html', $taxonomy_names ) )
						); ?>)</small>
			</label>
		</p>
		<p>
			<label for="<?php echo esc_attr( $this->get_field_id('custom_fields' ) ); ?>"><?php esc_html_e('Custom Fields:', 'mini-loops');?>
				<input class="widefat" id="<?php echo esc_attr( $this->get_field_id('custom_fields' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('custom_fields' ) ); ?>" type="text" value="<?php echo esc_attr( $custom_fields ); ?>" /><br />
				<small>(<?php esc_html_e('Ex: meta_key=meta_value&amp;meta_key2=meta_value2', 'mini-loops');?>)</small>
			</label>
		</p>
		<p>
			<label for="<?php echo esc_attr( $this->get_field_id('exclude' ) ); ?>"><?php esc_html_e('Exclude Posts:', 'mini-loops');?>
				<input class="widefat" id="<?php echo esc_attr( $this->get_field_id('exclude' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('exclude' ) ); ?>" type="text" value="<?php echo esc_attr( $exclude ); ?>" /><br />
				<small>(<?php esc_html_e('IDs, comma-separated.', 'mini-loops');?>)</small>
			</label>
		</p>
		<h3 style="clear:both;"><?php esc_html_e( 'Format', 'mini-loops' ); ?></h3>
		<p style="width:48%;float:left;">
			<label for="<?php echo esc_attr( $this->get_field_id('before_items' ) ); ?>"><?php esc_html_e('Before Item:', 'mini-loops');?>
				<?php /*<input class="widefat" id="<?php echo esc_attr( $this->get_field_id('before_items' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('before_items' ) ); ?>" type="text" value="<?php echo htmlspecialchars( stripslashes( $before_items ) ); ?>" />*/?>
				<textarea class="widefat" id="<?php echo esc_attr( $this->get_field_id('before_items' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('before_items' ) ); ?>"><?php echo esc_textarea( stripslashes( $before_items ) ); ?></textarea>
			</label>
		</p>
		<p style="width:48%;float:right;">
			<label for="<?php echo esc_attr( $this->get_field_id('after_items' ) ); ?>"><?php esc_html_e('After Item:', 'mini-loops');?>
				<?php /*<input class="widefat" id="<?php echo esc_attr( $this->get_field_id('after_items' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('after_items' ) ); ?>" type="text" value="<?php echo htmlspecialchars( stripslashes( $after_items ) ); ?>" />*/?>
				<textarea class="widefat" id="<?php echo esc_attr( $this->get_field_id('after_items' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('after_items' ) ); ?>"><?php echo esc_textarea( stripslashes( $after_items ) ); ?></textarea>
			</label>
		</p>
		<p>
			<label for="<?php echo esc_attr( $this->get_field_id('item_format' ) ); ?>"><?php esc_html_e('Item Format:', 'mini-loops');?>
				<textarea class="widefat" rows="5" id="<?php echo esc_attr( $this->get_field_id('item_format' ) ); ?>" name="<?php echo esc_attr( $this->get_field_name('item_format' ) ); ?>"><?php echo esc_textarea( $item_format ); ?></textarea>
			</label>
			<small><em><a href="http://wordpress.org/extend/plugins/mini-loops/other_notes/"><?php esc_html_e('See an explanation of options.', 'mini-loops');?></a></em></small>
		</p>

	</div>

	<div style="clear:both;"></div>
